### PR TITLE
internal/dag: Implement Gateway.Listener.Hostname when processing HTTPRoutes

### LINKED ISSUE #3561: Take the Gateway.Listener.hostname into account when processing routes
```
Hostname specifies the virtual hostname to match for protocol types that define this concept. When unspecified, “”, or *, all hostnames are matched. This field can be omitted for protocols that don’t require hostname based matching.

Hostname is the fully qualified domain name of a network host, as defined by RFC 3986. Note the following deviations from the “host” part of the URI as defined in the RFC:

IP literals are not allowed.
The : delimiter is not respected because ports are not allowed.
Hostname can be “precise” which is a domain name without the terminating dot of a network host (e.g. “foo.example.com”) or “wildcard”, which is a domain name prefixed with a single wildcard label (e.g. *.example.com). The wildcard character * must appear by itself as the first DNS label and matches only a single label.
```

//ref: https://gateway-api.sigs.k8s.io/spec/#networking.x-k8s.io/v1alpha1.Listener

### PR DESCRIPTION
**Pull request by stevesloka**
Take into account gateway.listeners.hostname into account when processing HTTPRoutes
which applies to all HTTPRoutes selected for that listener.

Fixes #3561

Signed-off-by: Steve Sloka <slokas@vmware.com>

### CODE REVIEW

File: internal/dag/builder_test.go [hunk 1]
```diff
@@ -2266,6 +2298,79 @@ func TestDAGInsertGatewayAPI(t *testing.T) {
...
+							"http.projectcontour.io",
+						},
+						Rules: []gatewayapi_v1alpha1.HTTPRouteRule{{
+							Matches:   httpRouteMatch(gatewayapi_v1alpha1.PathMatchExact, "/blog"),
+							ForwardTo: httpRouteForwardTo("kuard", 8080, 1),
+						}},
+					},
+				},
+			},
+			want: listeners(
```
Comment (sunjayBhatia) [inline internal/dag/builder_test.go:2362]: is there any documentation upstream about how we handle the route and listener specifying a hostname?

Comment (stevesloka) [reply internal/dag/builder_test.go:2362]: I assumed they were combined together which is what this PR does. I'm asking upstream to see if that's correct.

Comment (stevesloka) [reply internal/dag/builder_test.go:2362]: So my assumption is wrong, going to fix the docs upstream and rework this now. 

TLDR is that the listener `hostname` defines what the routes can use. So for a listener hostname of `*.test.com`, you could have a route with `foo.test.com`, but not `foo.steve.com`.

**Comment by stevesloka** [comment]
Ok I think I have this all fixed up now for how it should work. Going to PR a docs change upstream now to try and have a reference to validate against.

File: internal/dag/gatewayapi_processor.go [hunk 2]
```diff
@@ -195,39 +196,111 @@ func isSecretRef(certificateRef *gatewayapi_v1alpha1.LocalObjectReference) bool
...
+			} else if strings.Contains(hostname, "*") {
+
+				// Build a regex to match the hostname against the wildcard route hostname.
+				wildcardDNS1123SubdomainRegexp := regexp.MustCompile("^" + strings.Replace(hostname, "*", wildcardString, 1) + "$")
+
+				// Validate the wildcard route hostname matches gateway listener hostname.
+				if !wildcardDNS1123SubdomainRegexp.MatchString(lhn) {
+					errors = append(errors, fmt.Errorf("route hostname %q does not match gateway hostname %q", hostname, lhn))
+					continue
+				}
```
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:272]: can both the listener and route hostname be a wildcard, e.g. `*.foo.com` for both? If so then before doing the wildcard matching we should check for string equality between the listener and route host and add it if so.

File: internal/dag/gatewayapi_processor.go [hunk 3]
```diff
@@ -195,39 +196,111 @@ func isSecretRef(certificateRef *gatewayapi_v1alpha1.LocalObjectReference) bool
 	return strings.ToLower(certificateRef.Kind) == "secret" && strings.ToLower(certificateRef.Group) == "core"
 }
 
-func (p *GatewayAPIProcessor) computeHosts(route *gatewayapi_v1alpha1.HTTPRoute) ([]string, []error) {
-	// Determine the hosts on the route, if no hosts
-	// are defined, then set to "*".
+func (p *GatewayAPIProcessor) computeHosts(route *gatewayapi_v1alpha1.HTTPRoute, listenerHostname *gatewayapi_v1alpha1.Hostname) ([]string, []error) {
```
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:199]: A comment explaining the logic of this method would be helpful, including the details around how the route and listener hostnames relate.

**Review by skriss — COMMENTED** [review]
code looks pretty good, I had one question on the logic. Looking at tests next.

File: internal/dag/gatewayapi_processor_test.go [hunk 4]
```diff
@@ -26,11 +26,13 @@ import (
 
 func TestComputeHosts(t *testing.T) {
```
Comment (skriss) [inline internal/dag/gatewayapi_processor_test.go:27]: I think you could change `computeHosts` to not take a full `HTTPRoute`, just the hostnames string slice. It'd make the test setup easier here.

File: internal/dag/gatewayapi_processor_test.go [hunk 5]
```diff
@@ -184,9 +193,199 @@ func TestComputeHosts(t *testing.T) {
...
+				},
+				Spec: gatewayapi_v1alpha1.HTTPRouteSpec{
+					Hostnames: []gatewayapi_v1alpha1.Hostname{
+						"http.example.com",
+					},
+				},
+			},
+			want:      nil,
+			wantError: []error{fmt.Errorf("gateway hostname \"*.projectcontour.io\" does not match route hostname \"http.example.com\"")},
+		},
```
Comment (skriss) [inline internal/dag/gatewayapi_processor_test.go:312]: it'd be nice to add some cases to verify that e.g. `too.many.labels.projectcontour.io` doesn't match `*.projectcontour.io` (in both directions, i.e. where the listener has the wildcard and where the route has the wildcard) to get better coverage around the regex.

File: internal/dag/gatewayapi_processor_test.go [hunk 6]
```diff
@@ -184,9 +193,199 @@ func TestComputeHosts(t *testing.T) {
...
+				},
+				Spec: gatewayapi_v1alpha1.HTTPRouteSpec{
+					Hostnames: []gatewayapi_v1alpha1.Hostname{
+						"http.projectcontour.io",
+					},
+				},
+			},
+			want:      []string{"http.projectcontour.io"},
+			wantError: nil,
+		},
```
Comment (skriss) [inline internal/dag/gatewayapi_processor_test.go:292]: per previous comment, let's add a test case for when both the listener and route have e.g. `*.projectcontour.io` (not exactly sure what expected behavior is but let's define it via test).

File: internal/dag/gatewayapi_processor_test.go (hunk 5 above)
Comment (sunjayBhatia) [reply internal/dag/gatewayapi_processor_test.go:312]: yeah good point, probably need an integration test as well for wildcards if we're intending on supporting it already

File: internal/dag/gatewayapi_processor.go (hunk 3 above)
Comment (stevesloka) [reply internal/dag/gatewayapi_processor.go:199]: 👍

File: internal/dag/gatewayapi_processor.go (hunk 2 above)
Comment (stevesloka) [reply internal/dag/gatewayapi_processor.go:272]: Yup nice catch, I fixed that up and added a test for it.

File: internal/dag/gatewayapi_processor_test.go (hunk 4 above)
Comment (stevesloka) [reply internal/dag/gatewayapi_processor_test.go:27]: 👍

File: internal/dag/gatewayapi_processor_test.go (hunk 5 above)
Comment (stevesloka) [reply internal/dag/gatewayapi_processor_test.go:312]: I added some tests, I'll make some additional integration tests in a future PR.

**Comment by stevesloka** [comment]
Thanks for the comments @skriss @sunjayBhatia, I've got this updated.

File: internal/dag/accessors.go [hunk 7]
```diff
@@ -24,7 +24,7 @@ import (
 	"k8s.io/apimachinery/pkg/util/intstr"
 )
 
-// RouteServiceName identifies a service used in a route.
+// RouteServiceName identifies a service used in a hostnames.
```
Comment (skriss) [inline internal/dag/accessors.go:27]: probably an inadvertent find/replace?

File: internal/dag/conditions.go [hunk 8]
```diff
@@ -39,7 +39,7 @@ func mergePathMatchConditions(conds []contour_api_v1.MatchCondition) MatchCondit
 	// we need to set the prefix to /.
 	// Remember that this step is done AFTER all the includes have happened.
 	// Setting this to / allows us to pass this prefix to Envoy, as there must
-	// be at least one path, prefix, or regex set on each Envoy route.
+	// be at least one path, prefix, or regex set on each Envoy hostnames.
```
Comment (skriss) [inline internal/dag/conditions.go:42]: this file's changes also look like inadvertent find/replace

File: internal/dag/dag.go [hunk 9]
```diff
@@ -187,13 +187,13 @@ func (hc *HeaderMatchCondition) String() string {
 }
 
 // DirectResponse allows for a specific HTTP status code
-// to be the response to a route request vs routing to
+// to be the response to a hostnames request vs routing to
```
Comment (skriss) [inline internal/dag/dag.go:190]: this file too

File: internal/dag/gatewayapi_processor.go [hunk 10]
```diff
@@ -119,7 +120,7 @@ func (p *GatewayAPIProcessor) Run(dag *DAG, source *KubernetesCache) {
 
 			// ## RouteBindingSelector ##
 			//
-			// Selector specifies a set of route labels used for selecting routes to associate
+			// Selector specifies a set of hostnames labels used for selecting routes to associate
```
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:123]: looks like there are a bunch more find/replace errors in this and other file, I'm not gonna comment on them all, hopefully you can go through and revert all the erroneous ones.

File: internal/dag/gatewayapi_processor.go [hunk 11]
```diff
@@ -195,39 +196,121 @@ func isSecretRef(certificateRef *gatewayapi_v1alpha1.LocalObjectReference) bool
...
+				wildcardDNS1123SubdomainRegexp := regexp.MustCompile("^" + strings.Replace(hostname, "*", wildcardString, 1) + "$")
+
+				// Validate the wildcard hostnames hostname matches gateway listener hostname.
+				if !wildcardDNS1123SubdomainRegexp.MatchString(lhn) {
+					errors = append(errors, fmt.Errorf("hostnames hostname %q does not match gateway hostname %q", hostname, lhn))
+					continue
+				}
+			} else {
+				// Validate the gateway listener hostname matches the hostnames hostname.
+				if lhn != hostname {
```
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:285]: I think this check for non-equality is redundant/can be removed since we already checked for equality on https://github.com/projectcontour/contour/pull/3640/files#diff-2dfb3932d84ad36571b20cb8d66d9f0b623fdbbef49a0b41eff48eaa7fa22a05R258

File: internal/dag/gatewayapi_processor.go [hunk 12]
```diff
@@ -195,39 +196,121 @@ func isSecretRef(certificateRef *gatewayapi_v1alpha1.LocalObjectReference) bool
...
-				errors = append(errors, fmt.Errorf("invalid listener hostname %q: %v", hostname, errs))
+
+			// If the listener.hostname matches then no need to
+			// do any other validation.
+			if lhn == hostname {
+				hosts = append(hosts, hostname)
 				continue
 			}
+
+			if strings.Contains(lhn, "*") {
```
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:263]: nit: I would switch this to an `else if` joined with the above `if` since they all go together, for clarity, or could do a switch.

**Review by skriss — CHANGES_REQUESTED** [review]
Looks pretty good, just a couple minor style nits and some over-eager find/replaces to revert.

File: internal/dag/gatewayapi_processor.go (hunk 12 above)
Comment (stevesloka) [reply internal/dag/gatewayapi_processor.go:263]: They all have `continues` in them which showed they exit, but can update.

File: internal/dag/gatewayapi_processor.go (hunk 11 above)
Comment (stevesloka) [reply internal/dag/gatewayapi_processor.go:285]: We need this to cover this test case where the hostnames don't match without wildcards: https://github.com/projectcontour/contour/blob/d0ed425ce3a3d6950e989e3ae58963fea5e2bfae/internal/dag/gatewayapi_processor_test.go#L119

**Comment by stevesloka** [comment]
@skriss sorry about those renames, Goland's refactor+rename bit me again.

File: internal/dag/gatewayapi_processor.go (hunk 11 above)
Comment (skriss) [reply internal/dag/gatewayapi_processor.go:281]: We need the `else`, but since we already checked for `if lhn == hostname` above, then we know if we got here that `lhn != hostname`. It's not a big deal, can leave.

**Review by skriss — APPROVED** [review]
LGTM

File: internal/dag/gatewayapi_processor.go (hunk 11 above)
Comment (stevesloka) [reply internal/dag/gatewayapi_processor.go]: Ohh I see now, I thought you meant to remove the `if` statement. I updated. =)

**Review by skriss — APPROVED** [review]

File: internal/dag/gatewayapi_processor_test.go [hunk 13]
```diff
@@ -17,175 +17,190 @@ import (
...
+		"listener host & hostnames host do not match wildcard host": {
+			listenerHost: listenerHostname("*.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"http.example.com",
 			},
+			want:      nil,
+			wantError: []error{fmt.Errorf("gateway hostname \"*.projectcontour.io\" does not match hostnames hostname \"http.example.com\"")},
+		},
+		"listener host & wildcard hostnames host match": {
```
Comment (sunjayBhatia) [inline internal/dag/gatewayapi_processor_test.go:151]: is this allowed? this seems like the listener is more specific than the route, which could be not allowed?

File: internal/dag/gatewayapi_processor_test.go [hunk 14]
```diff
@@ -17,175 +17,190 @@ import (
...
+		},
+		"listener host wildcard & wildcard hostnames host match": {
+			listenerHost: listenerHostname("*.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"*.projectcontour.io",
+			},
+			want:      []string{"*.projectcontour.io"},
+			wantError: nil,
+		},
+		"listener host & wildcard all hostnames host match": {
```
Comment (sunjayBhatia) [inline internal/dag/gatewayapi_processor_test.go:167]: similar to the above, seems like this is a more specific gateway hostname than the route hostname

File: internal/dag/gatewayapi_processor.go [hunk 15]
```diff
@@ -195,39 +196,115 @@ func isSecretRef(certificateRef *gatewayapi_v1alpha1.LocalObjectReference) bool
...
+	}
+
+	for _, host := range hostnames {
 
 		hostname := string(host)
-		if isIP := net.ParseIP(hostname) != nil; isIP {
-			errors = append(errors, fmt.Errorf("hostname %q must be a DNS name, not an IP address", hostname))
+
+		// A "*" hostname matches anything.
+		if hostname == "*" {
```
Comment (sunjayBhatia) [inline internal/dag/gatewayapi_processor.go:235]: a `*` should only be able to come from line https://github.com/projectcontour/contour/pull/3640/files#diff-2dfb3932d84ad36571b20cb8d66d9f0b623fdbbef49a0b41eff48eaa7fa22a05R211 above, but could the order of this processing mean that we now allow `*` in a hostname on a route? (which doesn't seem like it should be allowed, per the docs)

File: internal/dag/gatewayapi_processor_test.go [hunk 16]
```diff
@@ -17,175 +17,190 @@ import (
...
+			listenerHost: listenerHostname("http.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"*",
```
Comment (sunjayBhatia) [inline internal/dag/gatewayapi_processor_test.go:170]: i think this is an illegal hostname on a route:

```
The wildcard character * must appear by itself as the first DNS label and matches only a single label. You cannot have a wildcard label by itself (e.g. Host == *).
```

see https://gateway-api.sigs.k8s.io/references/spec/#networking.x-k8s.io/v1alpha1.HTTPRoute hostname field

**Review by sunjayBhatia — COMMENTED** [review]
also my guess is that an integration test will show we need something similar to https://github.com/projectcontour/contour/blob/dee66310416af9e0477705916421a7e9a7219f99/internal/dag/ingress_processor.go#L233-L248

to make sure wildcard matching works as expected

File: internal/dag/gatewayapi_processor.go (hunk 15 above)
Comment (stevesloka) [reply internal/dag/gatewayapi_processor.go:235]: You're correct based on the docs. We can set a `*` for the hostname and not setting is the same thing, but agree should update.

File: internal/dag/gatewayapi_processor_test.go (hunk 14 above)
Comment (stevesloka) [reply internal/dag/gatewayapi_processor_test.go:167]: I guess this test would be the same as (https://github.com/projectcontour/contour/blob/d0ed425ce3a3d6950e989e3ae58963fea5e2bfae/internal/dag/gatewayapi_processor_test.go#L183) which essentially has no hostname defined on the Route, but the Gateway does, so it should end up only matching the Gateway.

File: internal/dag/gatewayapi_processor_test.go (hunk 13 above)
Comment (stevesloka) [reply internal/dag/gatewayapi_processor_test.go:151]: I'm not 100% sure, but the test is wrong. If the Gateway defines `hostname.projectcontour.io` and the route defines `*.projectcontour.io`, then I think `hostname.projectcontour.io` should end up being the match. Does that make sense?

**Comment by stevesloka** [comment]
Need to also add some tests thinking about this some more if the route defines multiple hostnames or the set of routes end up with more than one hostname.

**Comment by stevesloka** [comment]
I think this is ready for a final look @sunjayBhatia. I added the wildcard tests/changes to https://github.com/projectcontour/contour/pull/3701

File: internal/dag/gatewayapi_processor_test.go (hunk 13 above)
Comment (sunjayBhatia) [reply internal/dag/gatewayapi_processor_test.go:186]: yeah i think if we do allow this, the route programmed should be for the more specific one

I guess the question is can a route have a wildcard on a gateway/listener with a specific match? `*.projectcontour.io` is a superset of `http.projectcontour.io` so it is a bit odd to me this could be valid? idk

Comment (sunjayBhatia) [reply internal/dag/gatewayapi_processor_test.go:186]: cc @projectcontour/maintainers if there are any other opinions on this

File: internal/dag/gatewayapi_processor_test.go [hunk 17]
```diff
@@ -27,166 +27,251 @@ import (
...
+			listenerHost: listenerHostname("http.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"*.projectcontour.io",
+			},
+			want: map[string]struct{}{
+				"http.projectcontour.io": {},
+			},
+			wantError: nil,
+		},
+		"listener host & wildcard hostname and matching hostname match": {
```
Comment (sunjayBhatia) [inline internal/dag/gatewayapi_processor_test.go:196]: similar to the above comment on this one

File: internal/dag/gatewayapi_processor_test.go [hunk 18]
```diff
@@ -27,166 +27,251 @@ import (
...
+			want:      map[string]struct{}{},
+			wantError: []error{fmt.Errorf("hostnames hostname \"*.projectcontour.io\" does not match gateway hostname \"too.many.labels.projectcontour.io\"")},
+		},
+		"listener wildcard host doesn't match hostnames with many labels host": {
+			listenerHost: listenerHostname("*.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"too.many.labels.projectcontour.io",
+			},
+			want:      map[string]struct{}{},
+			wantError: []error{fmt.Errorf("gateway hostname \"*.projectcontour.io\" does not match hostnames hostname \"too.many.labels.projectcontour.io\"")},
```
Comment (sunjayBhatia) [inline internal/dag/gatewayapi_processor_test.go:274]: maybe add a test or two where a route tries to define a hostname `*` with the listener also allowing `*`

File: internal/dag/gatewayapi_processor_test.go (hunk 17 above)
Comment (stevesloka) [reply internal/dag/gatewayapi_processor_test.go:196]: My understanding is the listener.hostname defines the rules, so if it specifies a wildcard, and the route's hostname matches, then that route hostname is allowed to serve.

File: internal/dag/gatewayapi_processor_test.go (hunk 18 above)
Comment (stevesloka) [reply internal/dag/gatewayapi_processor_test.go:274]: ok added a test with `*` on listener with `*.projectcontour.io` on route.

**Comment by stevesloka** [comment]
I ended up squashing stuff since the rebase was horrible, sorry @sunjayBhatia

File: internal/dag/gatewayapi_processor_test.go (hunk 13 above)
Comment (danehans) [reply internal/dag/gatewayapi_processor_test.go:186]: @stevesloka IMO a wildcard hostname should not be allowed for a route bound to a gateway listener with a more precise hostname. The API does not specify this use case, so I suggest updating https://github.com/kubernetes-sigs/gateway-api/issues/654 to get input from the community.

File: internal/dag/builder_test.go [hunk 19]
```diff
@@ -2515,6 +2547,77 @@ func TestDAGInsertGatewayAPI(t *testing.T) {
...
+				&Listener{
+					Port: 80,
+					VirtualHosts: virtualhosts(
+						virtualhost("*.projectcontour.io",
+							exactrouteHTTPRoute("/blog", service(kuardService))),
+					),
```
[the remark this answers is not in the document]
Comment (sunjayBhatia) [reply internal/dag/builder_test.go:2646]: ok looks like this is resolved since we made the route use a precise hostname match with the wildcard gateway

File: internal/dag/gatewayapi_processor.go [hunk 20]
```diff
@@ -230,24 +230,98 @@ func isSecretRef(certificateRef *gatewayapi_v1alpha1.LocalObjectReference) bool
...
+				continue
+			} else if lhn == hostname {
+				// If the listener.hostname matches then no need to
+				// do any other validation.
+				hosts[hostname] = struct{}{}
+				continue
+			} else if strings.Contains(lhn, "*") {
+
+				// Build a regex to match the hostname against the wildcard listener hostname.
+				wildcardDNS1123SubdomainRegexp := regexp.MustCompile("^" + strings.Replace(lhn, "*", wildcardString, 1) + "$")
```
Comment (sunjayBhatia) [inline internal/dag/gatewayapi_processor.go:299]: nit: I think this could be simplified to not need a regex match? If the listener hostname is a wildcard (`*.foo.com`) just chop the first DNS label from that and the route hostname and those suffixes must be equal

not super critical but makes it harder to make a mistake in case we get our regex wrong etc.

File: internal/dag/gatewayapi_processor.go [hunk 21]
```diff
@@ -230,24 +230,98 @@ func isSecretRef(certificateRef *gatewayapi_v1alpha1.LocalObjectReference) bool
...
+
+		// Validate the hostname.
 		if err := validHostName(hostname); err != nil {
 			errors = append(errors, err)
 			continue
 		}
-		hosts = append(hosts, string(host))
+
+		if listenerHostname != nil {
+			const wildcardString string = "[a-z0-9]([-a-z0-9]*[a-z0-9])*."
```
Comment (sunjayBhatia) [inline internal/dag/gatewayapi_processor.go:284]: just noticed, do we have to escape the `.` at the end b/c we want a literal `.`?

File: internal/dag/gatewayapi_processor_test.go (hunk 13 above)
Comment (sunjayBhatia) [reply internal/dag/gatewayapi_processor_test.go:186]: commented on the upstream issue, not sure what we want to do for now, restrict to something less permissive and expand or open it up now and restrict later

File: internal/dag/gatewayapi_processor_test.go (hunk 18 above)
Comment (sunjayBhatia) [reply internal/dag/gatewayapi_processor_test.go:274]: sorry to be a pain but I think a test where the HTTPRoute tries to use `*` as a hostname should be added too, since that is invalid

I don't remember if somewhere else in the code already restricts us from using that but if so would be good to have an explicit test somewhere, here or in the builder or something?

File: internal/dag/gatewayapi_processor.go (hunk 20 above)
Comment (stevesloka) [reply internal/dag/gatewayapi_processor.go:299]: nope I like it!

File: internal/dag/gatewayapi_processor.go (hunk 21 above)
Comment (stevesloka) [reply internal/dag/gatewayapi_processor.go]: took out all the regex stuff with your suggestion.

**Comment by stevesloka** [comment]
> I don't remember if somewhere else in the code already restricts us from using that but if so would be good to have an explicit test somewhere, here or in the builder or something?

I added a test to be explicit, but it's checked here: https://github.com/projectcontour/contour/blob/af1856edc34c7423b05796d33d0f8063727225f5/internal/dag/gatewayapi_processor.go#L278

File: internal/dag/gatewayapi_processor.go [hunk 22]
```diff
@@ -233,28 +233,86 @@ func isSecretRef(certificateRef *gatewayapi_v1alpha1.LocalObjectReference) bool
...
+				continue
+			}
+		}
+		hosts[hostname] = struct{}{}
 	}
 	return hosts, errors
 }
 
+func removeFirstDNSLabel(input string) string {
+	return input[strings.IndexAny(input, ".")+1:]
```
Comment (sunjayBhatia) [inline internal/dag/gatewayapi_processor.go:313]: i think the one thing to be careful about with this is this case:

```
package main

import (
	"fmt"
	"strings"
)

func main() {
	input := "*.foo"
	fmt.Println(input[strings.IndexAny(input, ".")+1:])
	input = "foo"
	fmt.Println(input[strings.IndexAny(input, ".")+1:])
}
```

these both end up being `foo`

so to be safe, just return the whole string if it has no `.` and keep the `.` if it does?

Comment (stevesloka) [reply internal/dag/gatewayapi_processor.go]: nope makes sense!

File: internal/dag/gatewayapi_processor.go [hunk 23]
```diff
@@ -233,28 +233,89 @@ func isSecretRef(certificateRef *gatewayapi_v1alpha1.LocalObjectReference) bool
...
+				return hosts, nil
+			}
+		}
+	}
+
 	for _, host := range hostnames {
 
 		hostname := string(host)
+
+		//// A "*" hostname matches anything.
```
Comment (sunjayBhatia) [inline internal/dag/gatewayapi_processor.go:271]: sorry last thing 😅  probably can delete

**Review by sunjayBhatia — APPROVED** [review]
